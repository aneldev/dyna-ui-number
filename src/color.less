.dyna-number {
	&.dyna-ui-field-wrapper-color- {
		&WHITE_BLACK {
		}
		&WHITE_RED {
		}
		&BLACK_WHITE {
		}
		&TRANSPARENT_WHITE {
			.dyna-ui-input-control-element {
				color: white;
				background-color: transparent;
			}
		}
		&ORANGE_WHITE {
		}
	}
}
