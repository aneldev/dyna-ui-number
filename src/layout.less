.dyna-number {
	.dyna-ui-field-wrapper-container {
		padding: 0 !important;;
		min-width: 150px;  // IE11 bug fix

		.dyna-ui-field-wrapper-control-container {
			top: 0 !important;
			height: 100%;

			.dn-number--input-content {
				height: 100%;
				display: flex;
				justify-content: center;

				.dn-number--value-button {
					display: flex;
					align-items: center;
					transition: transform 250ms ease-in-out;
					user-select: none;
					padding: 0 20px;

					.icon-container {
					}
				}

				input {
					border: 0;
					text-align: center;
					width: 60px;
					font-weight: bold;

					&:focus {
						outline: none;
					}
				}

			}

		}
	}

	&.dyna-ui-field-wrapper--mode- {
		&VIEW {
			.dn-number--value-button {
				transform: scale(0.01);

			}
		}
		&EDIT {
			.dn-number--value-button {
				transform: scale(1);
			}
		}
	}

	&.dyna-ui-field-wrapper--size- {
		&XSMALL {
			.dn-number--value-button {
				font-size: 24px;
				position: relative;
				top: 2px;
			}
		}
		&SMALL {
			.dn-number--value-button {
				font-size: 24px;
				position: relative;
				top: 2px;
			}
		}
		&MEDIUM {
			.dn-number--value-button {
				font-size: 30px;
				position: relative;
				top: 2px;
			}
		}
		&LARGE {
			.dn-number--value-button {
				font-size: 46px;
				position: relative;
				top: 2px;
			}
		}
	}
}
