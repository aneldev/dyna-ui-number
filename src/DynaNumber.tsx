import * as React from "react";
import {DynaFastClick} from "dyna-ui-fast-click";
import {DynaFieldWrapper, EMode, EColor, EStyle, ESize} from "dyna-ui-field-wrapper"
import {isNumber, round, roundToString} from "dyna-loops";

import {faIcon} from "./utils/faIcon";

import "./layout.less";
import "./color.less";

export {EColor, EStyle}

export interface IDynaNumberProps {
  style?: EStyle;
  mode?: EMode;
  color?: EColor;
  size?: ESize;
  name?: string;
  label?: TContent;
  required?: TContent;
  min?: number;
  max?: number;
  value: number;
  step?: number;
  precision?: number; // decimal
  validationMessage?: TContent;
  footer?: TContent;
  inputProps?: React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;
  onChange?: (name: string, value: number) => void;
}

export type TContent = string | JSX.Element;

export class DynaNumber extends React.Component<IDynaNumberProps> {
  static defaultProps: IDynaNumberProps = {
    mode: EMode.EDIT,
    style: EStyle.INLINE_ROUNDED,
    color: EColor.WHITE_BLACK,
    size: ESize.MEDIUM,
    name: '',
    label: null,
    min: null,
    max: null,
    value: 0,
    step: 1,
    precision: 0,
    validationMessage: null,
    inputProps: {},
    footer: null,
    onChange: (name: string, value: number) => undefined,
  };

  private handleChange(value: string) {
    const {name, precision, min, max, onChange} = this.props;
    if (!isNumber(value)) return;
    const n: number = round(Number(value), precision);
    if (min != null && n < min) return;
    if (max != null && n > max) return;
    onChange && onChange(name, n);
  }

  private handleButtonValueClick(factor: number): void {
    const {name, step, value, min, max, onChange} = this.props;
    const newValue: number = value + (factor * step);
    if (min != null && newValue < min) return;
    if (max != null && newValue > max) return;
    onChange(name, newValue);
  }

  private renderValueButton(factor: number): JSX.Element {
    const icon: string = factor === +1 ? 'plus-circle' : 'minus-circle';

    return (
      <DynaFastClick
        className="dn-number--value-button"
        nodeType="div"
        onClick={this.handleButtonValueClick.bind(this, factor)}
      >{faIcon(icon)}</DynaFastClick>
    );
  }

  private getFormatedValue(): string {
    const {value, precision} = this.props;
    return roundToString(value, precision);
  }

  public render(): JSX.Element {
    const {
      style, color, mode, size,
      label, required,
      inputProps,
      validationMessage, footer,
    } = this.props;

    return (
      <DynaFieldWrapper
        className="dyna-number"
        mode={mode}
        style={style}
        color={color}
        size={size}
        inputElementSelector="input"
        label={label}
        required={required}
        validationMessage={validationMessage}
        footer={footer}
      >
        <div className="dn-number--input-content">
          {this.renderValueButton(-1)}
          <input
            readOnly
            value={this.getFormatedValue()}
            {...inputProps}
            onChange={e => this.handleChange(e.target.value)}
          />
          {this.renderValueButton(+1)}
        </div>
      </DynaFieldWrapper>
    );
  }
}
