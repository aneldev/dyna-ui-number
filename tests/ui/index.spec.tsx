declare let global: any, jasmine: any, describe: any, clearTest: any, it: any, expect: any;

import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import * as React from 'react';
import * as enzyme from 'enzyme';

import {DynaNumber} from '../../src';

describe('Home', () => {
  let wrapper;

  it('has expected content with deep render', () => {
    wrapper = enzyme.shallow(
      (
        <DynaNumber value={2} />
      ),
      {}
    );

    expect(wrapper).toMatchSnapshot();
  });
});
