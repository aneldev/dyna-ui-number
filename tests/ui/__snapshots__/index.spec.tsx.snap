// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home has expected content with deep render 1`] = `
<DynaFieldWrapper
  className="dyna-number"
  color="WHITE_BLACK"
  footer={null}
  inputElementSelector="input"
  isLoading={null}
  label={null}
  mode="EDIT"
  onClick={[Function]}
  onFocus={[Function]}
  onOutsideClick={[Function]}
  required={null}
  size="MEDIUM"
  style="INLINE_ROUNDED"
  validationMessage={null}
>
  <div
    className="dn-number--input-content"
  >
    <div
      className="dn-number--value-button"
      onClick={[Function]}
    >
      <div
        className="icon-container"
      >
        <i
          aria-hidden="true"
          className="fa fa-plus-circle"
        />
      </div>
    </div>
    <input
      onChange={[Function]}
      readOnly={true}
      value="2"
    />
    <div
      className="dn-number--value-button"
      onClick={[Function]}
    >
      <div
        className="icon-container"
      >
        <i
          aria-hidden="true"
          className="fa fa-minus-circle"
        />
      </div>
    </div>
  </div>
</DynaFieldWrapper>
`;
